### ISSUE TITLE
Restore "Unitary Synthesis of ChoiMixTableau for Diagonalisation" and fix regression

### ISSUE THREAD

**Issue by cqc-alec** [CQCL/tket#1202]
#1201 reverted #941 because it caused some regressions in compilation. For example, with this circuit and sequence of passes:

```python
import json
from pytket.circuit import Circuit, OpType
from pytket.passes import (
    DecomposeBoxes,
    FullPeepholeOptimise,
    auto_rebase_pass,
    auto_squash_pass,
    NormaliseTK2,
    DecomposeTK2,
    RemoveRedundancies,
)

circ_json_s = '{"bits": [["c", [0]]], "commands": [{"args": [["q", [0]]], "op": {"type": "H"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "4279cff1-f6da-495d-85ae-0db0ed5507c6", "paulis": ["I", "Z", "I"], "phase": "-0.0770387583087794", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "2b2fbc8a-9f64-46b8-84dc-a452dc7b1160", "paulis": ["Z", "Z", "I"], "phase": "0.0770387583087794", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "979f1e38-03cb-4bcc-903b-c0c7a341408a", "paulis": ["I", "X", "Z"], "phase": "-0.00635529488517404", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "45b2adb0-30b1-4921-a041-0c24d2e2bfc0", "paulis": ["Z", "X", "Z"], "phase": "0.00635529488517404", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "82c67b89-4918-47bc-8c2a-73ab8c2d9015", "paulis": ["I", "I", "Z"], "phase": "-0.0770387583087794", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "2f3643b5-970e-48a5-9a1c-f7fecdf2360e", "paulis": ["Z", "I", "Z"], "phase": "0.0770387583087794", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "3442de73-9654-464d-969f-50d923671c71", "paulis": ["I", "Z", "X"], "phase": "0.00635529488517404", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "e239f42d-f046-4db7-b1d9-407ee5ceab13", "paulis": ["Z", "Z", "X"], "phase": "-0.00635529488517404", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "d43ffb33-08b4-4076-915c-8498ed608440", "paulis": ["I", "Z", "Z"], "phase": "0.00204224797013461", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "1f5e14e2-7592-4121-accb-5ed3c17a7c43", "paulis": ["Z", "Z", "Z"], "phase": "-0.00204224797013461", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "e95796b9-4b0b-46c2-bc0f-665da19d0fbe", "paulis": ["I", "X", "I"], "phase": "0.00230162576968647", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "ea557424-9fde-4e2c-a632-e7db20ad8b63", "paulis": ["Z", "X", "I"], "phase": "-0.00230162576968647", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "1aa60a48-de3b-42ca-8ee2-71ea7fad2520", "paulis": ["I", "I", "X"], "phase": "-0.00230162576968647", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "33588eb7-44c5-44b8-bf5e-0059f4b607aa", "paulis": ["Z", "I", "X"], "phase": "0.00230162576968647", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "2fb205ac-7893-4dab-95d7-25b6b64a4827", "paulis": ["I", "Y", "Y"], "phase": "-0.0303544146850987", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "362d5fe0-f07f-4c01-9a7e-8463248fe3df", "paulis": ["Z", "Y", "Y"], "phase": "0.0303544146850987", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "48a81702-15c7-44c3-8415-ed7df9d6662f", "paulis": ["I", "I", "I"], "phase": "-0.0429859635195463", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]]], "op": {"box": {"cx_config": "Tree", "id": "c700fa92-0a92-437a-86ad-8bd20354e53c", "paulis": ["Z", "I", "I"], "phase": "0.0429859635195463", "type": "PauliExpBox"}, "type": "PauliExpBox"}}, {"args": [["q", [0]]], "op": {"type": "H"}}, {"args": [["q", [0]], ["c", [0]]], "op": {"type": "Measure"}}], "created_qubits": [], "discarded_qubits": [], "implicit_permutation": [[["q", [0]], ["q", [0]]], [["q", [1]], ["q", [1]]], [["q", [2]], ["q", [2]]]], "phase": "0.0", "qubits": [["q", [0]], ["q", [1]], ["q", [2]]]}'
circ = Circuit.from_dict(json.loads(circ_json_s))

rebase = auto_rebase_pass({OpType.PhasedX, OpType.Rz, OpType.ZZPhase})
squash = auto_squash_pass({OpType.PhasedX, OpType.Rz})

DecomposeBoxes().apply(circ)
FullPeepholeOptimise(target_2qb_gate=OpType.TK2).apply(circ)
NormaliseTK2().apply(circ)
DecomposeTK2(ZZPhase_fidelity=lambda x: 1.0).apply(circ)
rebase.apply(circ)
RemoveRedundancies().apply(circ)
squash.apply(circ)

print(circ.n_2qb_gates(), circ.depth())
```

(which is exactly what `QuantinuumBackend("H1-1E").default_compilation_pass(optimisation_level=2)` does), the 2-qubit gate count and depth went from 14 to 17 and 27 to 37  respectively.

With some sequences of passes, however, performance was improved: for example, in the above sequence if we just replace `DecomposeTK2(ZZPhase_fidelity=lambda x: 1.0)` with `DecomposeTK2()` we see an improvement (22 to 19 and 43 to 39 respectively). Ideally we'd want to keep those gains as well as fixing the regressions, for example by allowing alternative decompositions in `DecomposeBoxes()`.

**Comment by cqc-alec** [comment]
My analysis of the issue is that:

- After #941  `PauliExpBox`es were decomposed slightly differently from how they had been previously, with different sequences of CX gates.
- This can make a difference to how well `CliffordSimp()` then performs.

For example, with the following circuit:

```python
pbox0 = PauliExpBox([Pauli.I, Pauli.X, Pauli.Z], -0.25)
pbox1 = PauliExpBox([Pauli.Z, Pauli.X, Pauli.Z], 0.25)
pbox2 = PauliExpBox([Pauli.I, Pauli.Z, Pauli.X], 0.25)
circ = Circuit(3)
circ.add_pauliexpbox(pbox0, [0, 1, 2])
circ.add_pauliexpbox(pbox1, [0, 1, 2])
circ.add_pauliexpbox(pbox2, [0, 1, 2])
```

after applying `DecomposeBoxes()` we get a circuit containing the following subcircuit:

Pre- #941 :
```
... CX(2,1).H(1).H(1).CX(2,1).CX(1,0).Rz(0.25,0).CX(1,0).CX(2,1) ...
```
Post- #941 :
```
... CX(2,1).H(1).H(1).CX(1,0).CX(2,0).Rz(0.25,0).CX(2,0).CX(1,0) ...
```

These are equivalent; however, when `CliffordSimp()` is run on the first we get a depth-4 circuit with 3 CXs, whereas when it is run on the second we get a depth-6 circuit with 5 CXs.

Presumably two CX(2,1) gates were effectively cancelled when building the tableau.

So I think that the best way to fix this issue is to improve `CliffordSimp()`. Can we make it so that it gives good results for these two circuits?

**Comment by cqc-alec** [comment]
@willsimmons1465 I would be interested in any ideas you have regarding the comment above.

**Comment by willsimmons1465** [comment]
Apologies for the delayed response to this.

As you spotted, the top example is one where the two-qubit gates align poorly between successive boxes with the new decomposition, preventing peephole optimisations from eliminating them. What's weird about this example is that starting with `PauliSimp()` (the pass that is designed to exploit the flexibility in alignment of 2-qubit gates and ordering of Pauli exponentials in this kind of circuit) improves these stats initially but gives worse results after `FullPeepholeOptimise()`.

This speaks to the general unpredictability of changing the alignment of 2-qubit gates in decompositions that makes `PauliSimp()` somewhat unreliable and the choice of `PauliSimp()`, `ZXGraphlikeOptimisation()` and the similar methods of pyZX difficult to compare as there is no choice that always beats the others.

Offering alternative decompositions in `DecomposeBoxes()` could help, but the unpredictability of performance will be difficult to choose sensible defaults or convey to users how to find the one that works best for their circuits. It only really becomes simple once we can automatically run it with multiple of them and select the best outcome, optimising over the global circuit after subsequent compiler passes have been applied. The current API might also conflict with this change since the `CXConfigType` (fundamentally a parameter of synthesis) is set upon creation of the boxes and may be varied between boxes in the same circuit, so we may need to decouple this.

As for improving `CliffordSimp()`, I don't think it is really possible within the current remit of the pass. Its goal is to identify non-local applications of certain classes of Clifford patterns - notably ones that only involve a single pair of 2-qubit gates. The kind of analysis needed to identify a reduction in the final circuit involves more qubits, and has to be aware of the interaction between Clifford and non-Clifford gates. This seems much more to be in the remit of `PauliSimp()`, or another pass that operates at a higher logical level and understands Pauli exponentials as first-class objects and is intentionally trying to find good alignments.

Even though the changes to `PauliExpBox` synthesis in #941 were intended to fix an inconsistency in how we were using `CXConfigType` s, it seems like those inconsistencies and idiosyncrasies can be (at least situationally) useful. If we reintroduce the contents of #941, I'd be inclined to still move the PauliGadget synthesis code to Diagonalisation (adding them as specialised diagonalisation methods that only work for a certain number of strings), and make PauliGadget synthesis via calling diagonalisation methods with the existing synthesis as a default. This means we have both code solutions together in a structured way that would keep it reasonable for maintenance, and we preserve the existing behaviour exactly by default. Does this sound ok?

**Comment by cqc-alec** [comment]
Closed by #1320 .
